From: Mark Tinguely <tinguely@sgi.com>
Subject: dmapi: fix value from newer Linux strnlen_user()
References: bsc#932897
Patch-mainline: Never (dmapi isn't upstream)

Community commit v3.4-7800-ga08c535 implemented a generic
strnlen_user(). That function can return the length of the
user string but fail to restrict it to the supplied maximum.
This could over write uninitialized variable and possibly
the stack.

Manually restrict the string length by comparing the
strnlen_user() result to DM_SESSION_INFO_LEN.

Signed-off-by: Mark Tinguely <tinguely@sgi.com>
Reviewed-by: Ben Myers <bpm@sgi.com>
Acked-by: Jan Kara <jack@suse.cz>

---
 fs/dmapi/dmapi_session.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/fs/dmapi/dmapi_session.c
+++ b/fs/dmapi/dmapi_session.c
@@ -500,6 +500,9 @@ dm_create_session(
 	unsigned long		lc;		/* lock cookie */
 
 	len = strnlen_user(info, DM_SESSION_INFO_LEN-1);
+	/* must check length from strnlen_user since v3.4-7800-ga08c535 */
+	if (len > DM_SESSION_INFO_LEN)
+		len = DM_SESSION_INFO_LEN;
 	if (copy_from_user(sessinfo, info, len))
 		return(-EFAULT);
 	lc = mutex_spinlock(&dm_session_lock);
